File: Assignment 2: Syntax, Semantics, and Memory Management/Part2_Memory_Management/MemoryManagement.java
public class Main {
    public static void main(String[] args) {
        // Allocating memory dynamically
        String data = new String("Hello, Java!");

        // Garbage collector automatically handles unused memory
        processData(data);

        // Data is still accessible after being passed to a method
        System.out.println("Original data: " + data);

        // Nullifying reference to make the object eligible for garbage collection
        data = null;

        // Suggest garbage collection (not guaranteed)
        System.gc();
        System.out.println("Program completed.");
    }

    public static void processData(String data) {
        System.out.println("Processed data: " + data);
    }
}

File: Assignment 2: Syntax, Semantics, and Memory Management/Part2_Memory_Management/memory_management.rs

fn main() {
    // Ownership: Allocating memory on the heap
    let data = String::from("Hello, Rust!");

    // Borrowing: Passing reference to avoid ownership transfer
    print_data(&data);

    // The original owner still has access to the data
    println!("Original data: {}", data);

    // Ownership transfer: `data` is moved
    let moved_data = data;
    // println!("{}", data); // Error: data is no longer accessible

    println!("Moved data: {}", moved_data);
}

fn print_data(data: &String) {
    println!("Borrowed data: {}", data);
}

